### ISSUE TITLE
Addon doesn't work with text under Shadow DOM

### ISSUE THREAD

**Issue by peace2000** [fastaddons/AutoHighlight#6]
**Describe the bug**  
Text that's under Shadow DOM doesn't work properly. It gets highlighted for a while but then highlighting disappears.

**Steps To Reproduce**  
1. Open https://www.virustotal.com/
2. Try to highlight word "Analyze"

**Expected behavior**  
Highlighting would work without disappearing.

**Device info:**
 - operating system: Windows 10
 - Browser and version: FF 93.0
 - ScrollAnywhere version: -

**Additional info**
Add any other info about the problem here.

**Comment by fastaddons** [comment]
Interesting use case.
But I don't think I can fix this.
The actual text search is handled by a 3rd party library Mark.js:
https://github.com/julmot/mark.js/
My addon is only setting the styles and handling page-change detection.

But I'm not sure it can be fixed at all, the Shadow Root is encapsulating things inside. This is the first time I see it being used on a whole page.
The fact that it works for a second when page loads is probably a static HTML without shadow roots loaded by the service worker which is then replaced by live page... I think (some modern frameworks uses this technique because it's SEO friendly).

**Comment by peace2000** [comment]
Hi, thanks for replying.

Ok, though how come web browser's search function can find those texts?

**Comment by fastaddons** [comment]
Browser build-in features has elevated permissions, so they can access everything everywhere. 
Addons and scripts running in a page has very limited powers.

But I've just noticed the shadow root on the page is in "open" mode, so the script should be able to access the data inside, although only through the "shadowRoot" property. 
Then again, this is something that would have to be added into the "mark.js" library that I'm using. But the library wasn't updated in 3 years.
I will try to look into it, maybe I can try to upgraded it myself. Although I'm sure it won't be easy...

**Comment by peace2000** [comment]
I made a feature suggestion post to mark.js authors but don't know about their dev pace...

But apparently it's possible to do that, I was given this link: https://jsfiddle.net/zrnp0ugh/

**Comment by peace2000** [comment]
Hi, @fastaddons.

I'm afraid that changes to mark.js aren't going to be implemented https://github.com/julmot/mark.js/pull/450

However, I noted that you have added the version made by angezid as an option to AH:

![image](https://user-images.githubusercontent.com/17256841/194017107-72463fe3-c543-4249-8a89-4b74e6bebaf9.png)

Would it be possible to integrate this solution https://jsfiddle.net/zrnp0ugh/ as well to AH? I don't know if that demo can be directly added/implemented but decided to ask :) angezid made that demo last year in mark.js repo.

I'm using AH to highlight email fields in a ServiceNOW based website. I read lots of long email conversations in plaintext form and I use AH to highlight lines that have texts like "Sent:", "From:", "Subject:", "Date" etc to make plaintext email threads a lot more readable.

This addon works greatly for that purpose, but the issue is that ServiceNOW randomly wraps those threads under Shadow DOM. If that happens, highlighting doesn't work. (There is no clear pattern why it randomly puts them under Shadow DOM but sometimes not.)

**Comment by fastaddons** [comment]
Yeah, the original author seems to be too busy and the pull-request there is crazy huge. But the hope still exists.... The "Closed" state is only set by the automatic bot, not the author.

In any case, the jsfiddle seems simple enough, all he does is calling the highlight on the shadow root directly:
```
const root = document.querySelector('shadow-dom').shadowRoot;
new Mark(root.querySelector('p')).mark('text');
```
I could do that probably as well.
I will try to experiment with this in the next version.
But I'm currently "over my head" with things to do, so it may take a while.

**Comment by peace2000** [comment]
Thanks, I really appreciate!

**Comment by peace2000** [comment]
Hi again @fastaddons

I noticed you have switched `mark.js` to `advanced-mark.js`.

According to angezid, it should support texts under Shadow DOM, but I can't get it to work.

https://github.com/angezid/advanced-mark.js/issues/12

**Comment by Juraj-Masiar** [comment]
Done,
4.9 was just released for Firefox with Enable highlighting inside "Shadow DOM" option.
It works on Virustotal!
So far so good :)

Chrome / Edge version will be released later, they are still reviewing 4.8 which I submitted only this morning.

**Comment by peace2000** [comment]
@Juraj-Masiar 

Hi, tested this with Firefox and with Chrome. Works without issues.

Tested on https://www.virustotal.com/gui/home/upload
Also on nypost.com comment section: https://nypost.com/2023/04/20/merrick-garland-at-center-of-hunter-biden-whistleblower-claim/
And also on a ServiceNOW-based platform.

Thanks :)
